compiler/rustc_const_eval/src/interpret/eval_context.rs

```diff
@@ -896,7 +896,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 .local_to_op(self.frame(), mir::RETURN_PLACE, None)
                 .expect("return place should always be live");
             let dest = self.frame().return_place.clone();
-            let err = self.copy_op_allow_transmute(&op, &dest);
+            let err = if self.stack().len() == 1 {
+                // The initializer of constants and statics will get validated separately
+                // after the constant has been fully evaluated. While we could fall back to the default
+                // code path, that will cause -Zenforce-validity to cycle on static initializers.
+                // Reading from a static's memory is not allowed during its evaluation, and will always
+                // trigger a cycle error. Validation must read from the memory of the current item.
+                self.copy_op_no_validation(&op, &dest)
```

Comment:
@RalfJung this is a bit fishy, as it will also mean that miri won't validate the `_start` frame's return value. I can fix this, but it requires more code ^^

Reply:
Miri does `read_target_isize` on that return value, so it is being validated already.